pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -225,6 +227,9 @@ func (pl *DefaultPreemption) SelectVictimsOnNode(
 			return nil, 0, framework.AsStatus(err)
 		}
 	}
+
+	// Sort victims after reprieving pods to keep the victims sorted by pod priority from high to low.
+	sort.Slice(victims, func(i, j int) bool { return util.MoreImportantPod(victims[i], victims[j]) })
```

Comment:
Logic is fine, however, we can potentially optimize this to avoid unnecessary sorting:

1. thorough fix to remove sorting: in filterPodsWithPDBViolation(), return a custom struct to wrap PodInfo to carry a "isPDBViolated" flag so as to return a single slice; and hence we don't need two `for` loop, and hence no need to re-sort at the end of this function.
2. only do sorting when `len(violatingVictims) != 0 && len(nonViolatingVictims) != 0`.

cc @alculquicondor @sanposhiho

Reply:
ok, updated. Remove additional sorting and reduce time complexity from O(n)log(n) to O(n).

Please review, thanks. Feel free to comment if there are any better ways.

Reply:
Maybe another sorting is not a bad idea for cherry-pick? What do you guys think? And we still need to loop twice, no?

Reply:
> Maybe another sorting is not a bad idea for cherry-pick? What do you guys think? And we still need to loop twice, no?
Thanks review~ I'm ok for both impls. Current impl in MR is more elegant, it  drops sorting and uses one for loop instead, which reduces time complexity from O(n)log(n) to O(n). But add another sorting has less code intrusion，which benefits for understanding and cherry-pick...

Look forward for more good ideas or advices~

Reply:
The current impl. doesn't qualify as a cherry-pickable fix. Actually, it doesn't quite follow the proposal 1 in my previous [comment](https://github.com/kubernetes/kubernetes/pull/128307#discussion_r1818265050).

In proposal 1, I suggested to only change filterPodsWithPDBViolation() to return one slice, sth. like this:

```diff
diff --git a/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go b/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go
index f1c58de9389..a255aef74a8 100644
--- a/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go
+++ b/pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go
@@ -135,6 +135,11 @@ func (pl *DefaultPreemption) CandidatesToVictimsMap(candidates []preemption.Cand
 	return m
 }
 
+type potentialVictim struct {
+	podInfo     *framework.PodInfo
+	pdbViolated bool
+}
+
 // SelectVictimsOnNode finds minimum set of pods on the given node that should be preempted in order to make enough room
 // for "pod" to be scheduled.
 func (pl *DefaultPreemption) SelectVictimsOnNode(
@@ -195,7 +200,7 @@ func (pl *DefaultPreemption) SelectVictimsOnNode(
 	// Try to reprieve as many pods as possible. We first try to reprieve the PDB
 	// violating victims and then other non-violating ones. In both cases, we start
 	// from the highest priority victims.
-	violatingVictims, nonViolatingVictims := filterPodsWithPDBViolation(potentialVictims, pdbs)
+	finalVictims := filterPodsWithPDBViolation(potentialVictims, pdbs)
 	reprievePod := func(pi *framework.PodInfo) (bool, error) {
 		if err := addPod(pi); err != nil {
 			return false, err
@@ -212,19 +217,15 @@ func (pl *DefaultPreemption) SelectVictimsOnNode(
 		}
 		return fits, nil
 	}
-	for _, p := range violatingVictims {
-		if fits, err := reprievePod(p); err != nil {
+
+	for _, v := range finalVictims {
+		if fits, err := reprievePod(v.podInfo); err != nil {
 			return nil, 0, framework.AsStatus(err)
-		} else if !fits {
+		} else if !fits && v.pdbViolated {
 			numViolatingVictim++
 		}
 	}
-	// Now we try to reprieve non-violating victims.
-	for _, p := range nonViolatingVictims {
-		if _, err := reprievePod(p); err != nil {
-			return nil, 0, framework.AsStatus(err)
-		}
-	}
+
 	return victims, numViolatingVictim, framework.NewStatus(framework.Success)
 }
 
@@ -287,12 +288,13 @@ func podTerminatingByPreemption(p *v1.Pod) bool {
 // preempted.
 // This function is stable and does not change the order of received pods. So, if it
 // receives a sorted list, grouping will preserve the order of the input list.
-func filterPodsWithPDBViolation(podInfos []*framework.PodInfo, pdbs []*policy.PodDisruptionBudget) (violatingPodInfos, nonViolatingPodInfos []*framework.PodInfo) {
+func filterPodsWithPDBViolation(podInfos []*framework.PodInfo, pdbs []*policy.PodDisruptionBudget) []*potentialVictim {
 	pdbsAllowed := make([]int32, len(pdbs))
 	for i, pdb := range pdbs {
 		pdbsAllowed[i] = pdb.Status.DisruptionsAllowed
 	}
 
+	var potentialVictims []*potentialVictim
 	for _, podInfo := range podInfos {
 		pod := podInfo.Pod
 		pdbForPodIsViolated := false
@@ -326,13 +328,15 @@ func filterPodsWithPDBViolation(podInfos []*framework.PodInfo, pdbs []*policy.Po
 				}
 			}
 		}
+		v := &potentialVictim{
+			podInfo: podInfo,
+		}
 		if pdbForPodIsViolated {
-			violatingPodInfos = append(violatingPodInfos, podInfo)
-		} else {
-			nonViolatingPodInfos = append(nonViolatingPodInfos, podInfo)
+			v.pdbViolated = true
 		}
+		potentialVictims = append(potentialVictims, v)
 	}
-	return violatingPodInfos, nonViolatingPodInfos
+	return potentialVictims
 }
 
 func getPDBLister(informerFactory informers.SharedInformerFactory) policylisters.PodDisruptionBudgetLister {

```

Reply:
But we need to `reprieve the PDB violating victims` first, then non-PDB ones, what you suggest here only follows the priority ordering.

Reply:
Am I understanding right here?

Reply:
@Huang-Wei 
ahh, I got you. But I found this impl may reprieve non-violating pods instead of those violating pods. 

For example, there are two suitable pods podA(midPriority, nonViolatingPDB) & podB(lowPriority, violatingPDB) for scheduling podC(highPriority). Evicting any of them can make podC schedulable.

Clearly, we should preempt podB which violates PDB. But if we put podA & podB into one slice(finalVictims), we will reprieve podA firstly. Please correct me if I misunderstanding.

Reply:
> Current impl in MR is more elegant, it drops sorting and uses one for loop instead, which reduces time complexity from O(n)log(n) to O(n)

But the n is different, one refers to the final victims, another one refers to the potential victims.

Reply:
Yup, the above snippet was a rough idea. You may need to tweak a bit to adapt to the case to always favor the nodes w/ non-violating victims only.

Reply:
> > Current impl in MR is more elegant, it drops sorting and uses one for loop instead, which reduces time complexity from O(n)log(n) to O(n)
> 
> But the n is different, one refers to the final victims, another one refers to the potential victims.

yes, correct. Sorting for final victims with O(n1)log(n1), Another for all potential victims O(n2), and n1 <= n2.

Seems sorting is a good way for impl.

Reply:
Yup, let's go with sorting as a cherry-pickable fix. And revaluate a potential follow-up in main branch.